pkg/deploy/deployment.go

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package deploy
+
+import (
+	"context"
+	"fmt"
+
+	orgv1 "github.com/eclipse/che-operator/pkg/apis/org/v1"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	runtimeClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var deploymentDiffOpts = cmp.Options{
+	cmpopts.IgnoreFields(appsv1.Deployment{}, "TypeMeta", "ObjectMeta", "Status"),
+	cmpopts.IgnoreFields(appsv1.DeploymentSpec{}, "Replicas", "RevisionHistoryLimit", "ProgressDeadlineSeconds"),
+	cmpopts.IgnoreFields(appsv1.DeploymentStrategy{}, "RollingUpdate"),
+	cmpopts.IgnoreFields(corev1.Container{}, "TerminationMessagePath", "TerminationMessagePolicy"),
+	cmpopts.IgnoreFields(corev1.PodSpec{}, "DNSPolicy", "SchedulerName", "SecurityContext"),
+	cmpopts.IgnoreFields(corev1.VolumeSource{}, "EmptyDir"),
+	cmp.Comparer(func(x, y resource.Quantity) bool {
+		return x.Cmp(y) == 0
+	}),
+}
+
+type DeploymentProvisioningStatus struct {
+	ProvisioningStatus
+	Deployment *appsv1.Deployment
+}
+
+func SyncDeploymentToCluster(
+	checluster *orgv1.CheCluster,
+	specDeployment *appsv1.Deployment,
+	clusterDeployment *appsv1.Deployment,
+	customDeploymentDiffOpts cmp.Options,
+	customDeploymentMerge func(*appsv1.Deployment, *appsv1.Deployment) *appsv1.Deployment,
+	clusterAPI ClusterAPI) DeploymentProvisioningStatus {
+
+	clusterDeployment, err := getClusterDeployment(specDeployment.Name, specDeployment.Namespace, clusterAPI.Client)
+	if err != nil {
+		return DeploymentProvisioningStatus{
+			ProvisioningStatus: ProvisioningStatus{Err: err},
+		}
+	}
+
+	if clusterDeployment == nil {
+		logrus.Infof("Creating a new object: %s, name %s", specDeployment.Kind, specDeployment.Name)
+		err := clusterAPI.Client.Create(context.TODO(), specDeployment)
+		return DeploymentProvisioningStatus{
+			ProvisioningStatus: ProvisioningStatus{Requeue: true, Err: err},
+		}
+	}
+
+	if customDeploymentDiffOpts != nil {
```

Comment:
Not exactly.
`deploymentDiffOpts` exists for all deployments
`customDeploymentDiffOpts` it is an additional diff opts added for keycloak deployment.
I will rename it not to confuse.

Reply:
I understand the behavior now: in case of Keycloak deployment, you process the annotations first if necessary, and then the spec will be reconcile during the next round.

Maybe a small comment to explain this 2-step process would help ?

Reply:
Sure, I will add